Add an undoing & redoing system to the editor

AllyTally opened the pull request:
## Changes:

It's very easy to make mistakes in VVVVVV's built-in level editor, with no way to undo them. This PR adds an undo and redo system, bound to CTRL+Z for undo and CTRL+Y for redo. The undo and redo stacks don't have any limits, but could easily happen in the future. VVVVVV's data is small enough where this should never be an issue, however.

No notes show up for undoing and redoing, because keeping track of what specific action you're doing may bloat the system, and would get annoying for the end-user. Notes are a bit annoying in general, even, but that's not relevant for this PR.

## Legal Stuff:

By submitting this pull request, I confirm that...

- [x] My changes may be used in a future commercial release of VVVVVV
- [x] I will be credited in a `CONTRIBUTORS` file and the "GitHub Friends"
  section of the credits for all of said releases, but will NOT be compensated
  for these changes

AllyTally committed:
Add undoing & redoing to the editor

It's very easy to make mistakes in VVVVVV's built-in level editor, with
no way to undo them. This commit adds an undo and redo system, bound to
CTRL+Z for undo and CTRL+Y for redo. The undo and redo stacks don't
have any limits, but could easily happen in the future. VVVVVV's data
is small enough where this should never be an issue, however.

No notes show up for undoing and redoing, because keeping track of what
specific action you're doing may bloat the system, and would get
annoying for the end-user. Notes are a bit annoying in general, even.

AllyTally committed:
Add level resizing to undo/redo system

InfoTeddy commented:
Conflicts need to be resolved.

InfoTeddy commented on desktop_version/src/Editor.cpp:
```diff
@@ -3192,6 +3320,123 @@ void editorclass::get_input_line(const enum TextMode mode, const std::string& pr
+
+    if (buffer->size() == 0)
+    {
+        ed.show_note(loc::gettext(undo ? "ERROR: Nothing to undo" : "ERROR: Nothing to redo"));
```
In general we try to keep all `loc::gettext` strings easily greppable. Please move the `loc::gettext` calls into the ternary, not surrounding it.

Speaking of which, you need to add these strings to the language files too.

InfoTeddy reviewed (changes requested):
Style pass.

NyakoFox committed:
searchable gettext for undo/redo notes

NyakoFox committed:
A few fixes and cleanup

The main fix was how holding right click to remove tiles did not
commit an action.

NyakoFox committed:
Spikes also need to commit

NyakoFox committed:
Cancelling placing a terminal/script box shouldn't have undo events

NyakoFox committed:
Don't commit tile events if removing entities

InfoTeddy commented on desktop_version/src/Editor.cpp:
```diff
@@ -412,6 +412,8 @@ void editorclass::reset(void)
     state = EditorState_DRAW;
     substate = EditorSubState_MAIN;
+
+    undo_buffer.clear();
```
When a new level is loaded, the undo buffer is cleared but not the redo buffer. So if there's anything in the redo buffer, you can press Ctrl+Y and that will be applied in the new level when it shouldn't be.

NyakoFox committed:
Clear redo buffer on level load

NyakoFox committed:
Remove unneeded pointers from memcpy

NyakoFox committed:
Add strings to lang files

NyakoFox committed:
Fix definition in Editor.h

NyakoFox commented:
Rebased and force-pushed, no longer conflicts (and fixes an issue)

flibitijibibo commented on desktop_version/src/Editor.cpp:
```diff
@@ -3544,6 +3862,9 @@ void editorinput(void)
                     if (ed.old_entity_text == "")
                     {
                         ed.remove_entity(ed.text_entity);
+                        // Uncommit the past two actions
```
This could be improved with a 'why', in case someone looks at this and thinks it's just temporary kludge or something

NyakoFox replied:
Added a few comments to explain the logic, hopefully they makes sense!

flibitijibibo reviewed (approved):
At a high level this looks good - if anybody's tested this would be good to get a +1 from them as well before merging.

NyakoFox committed:
Explain double-uncommit

NyakoFox commented:
I asked a few people to test; seems like we're getting some weird memory corruption to do with cross-room undos, so I'm looking into that right now.

NyakoFox committed:
Don't unconditionally commit tiles on room change

I'm not that sure what I was thinking back then, but switching rooms
would commit tiles no matter what, causing a huge mess. The behavior
was intended to stop the "tile placement" action and commit it, but
just happened no matter what... So this has been fixed, and a few
comments have been added as well.

NyakoFox commented:
Fixed the issue; it was just a missed conditional somewhere which caused the stack to inflate & save tile states which didn't change and relied on old data (which made things work strangely) -- not the "memory corruption" I assumed it was. I'm waiting on people to test now and I'll report back once no issues have been found!

Daaaav commented on desktop_version/src/Editor.cpp:
```diff
@@ -2476,27 +2592,34 @@ void editorclass::entity_clicked(const int index)
     case 19:
         // Script Boxes (and terminals)
+        commit_entity(index);
         get_input_line(TEXT_SCRIPT, "Enter script name:", &entity->scriptname);
```
These two `get_input_line`s are inconsistent with each other - the new one adds a `loc::gettext` that the others didn't have (`get_input_line` is a rare case where the `loc::gettext` is done somewhere later)

NyakoFox replied:
Solved.

Daaaav commented on desktop_version/src/Editor.cpp:
```diff
@@ -3194,6 +3344,127 @@ void editorclass::get_input_line(const enum TextMode mode, const std::string& pr
     old_entity_text = key.keybuffer;
 }
 
+void process_editor_buffer(const bool undo)
```
`process_editor_buffer` is not a very clear name for a function that (I think) handles specifically the action of undoing or redoing.

NyakoFox replied:
Changed to `handle_undo`.

Daaaav reviewed (changes requested):
Few really small things, otherwise looks good I think!

NyakoFox committed:
Fix mentioned issues & make process_undo static

NyakoFox commented:
Fixed those; and I've gotten a few people to test again and they also say it looks good.
